webauthn-rs-core/src/fake.rs

```diff
@@ -0,0 +1,323 @@
+///
+/// Account enumeration defence is a very subjective matter. For example on a website for a
+/// psychologists or drug addiction service, account enumeration prevention is important for
+/// privacy of the users of the service. For a service such as a business it may be less important
+/// as corporate directories may be available.
+///
+/// Because of this some considerations have been made in this api.
+///
+/// The only input we have from the account enumeration is the username.
+///
+/// Two users of this library should not have the same generated credentials given the same username.
+///
+/// `CredentialID`s that are generated should be believable and appear to come from legitimate services
+/// or credential sources. This includes realistic distribution of the types of devices that users
+/// may be using.
+///
+/// This api must be consistent - the same inputs will yield the same outputs.
+///
+/// To achieve this, the generator uses a seeded CSPRNG for each operation. The CSPRNG is seeded
+/// from the HMAC of the username. The HMAC is keyed from an input that the site provides.
+///
+/// The HMAC key should not be disclosed, as knowledge of the HMAC key will allow an external
+/// party to determine which IDs are generated and which are not.
+pub struct WebauthnFakeCredentialGenerator<D>
+where
+    D: FakeCredentialIDDistribution,
+{
+    // hmac key
+    hmac_key: pkey::PKey<pkey::Private>,
+    distribution: std::marker::PhantomData<D>,
+}
+
+impl<D> WebauthnFakeCredentialGenerator<D>
+where
+    D: FakeCredentialIDDistribution,
+{
+    /// Create a new HMAC keyed fake credential generator. You should associate a distribution type
+    /// using type annotations.
+    pub fn new(hmac_key: &[u8]) -> Result<Self, WebauthnError> {
+        let hmac_key = pkey::PKey::hmac(hmac_key).map_err(WebauthnError::OpenSSLError)?;
+
+        Ok(WebauthnFakeCredentialGenerator {
+            hmac_key,
+            distribution: std::marker::PhantomData,
+        })
+    }
+
+    /// Given a username as a byte slice, generate a set of deterministic `CredentialID`s.
+    pub fn generate(&self, username: &[u8]) -> Result<Vec<CredentialID>, WebauthnError> {
+        // hmac the username
+        let mut signer = sign::Signer::new(hash::MessageDigest::sha256(), &self.hmac_key)
+            .map_err(WebauthnError::OpenSSLError)?;
+
+        let mut seed = [0; 32];
+        let buf = signer
+            .sign_oneshot_to_vec(username)
+            .map_err(WebauthnError::OpenSSLError)?;
+
+        seed.copy_from_slice(&buf);
+
+        // Seed the rng
+        let mut seeded_rng = ChaCha8Rng::from_seed(seed);
+
+        let credentials = D::generate(&mut seeded_rng);
+
+        Ok(credentials)
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::{FakePasskeyDistribution, WebauthnFakeCredentialGenerator};
+    use crate::proto::Base64UrlSafeData;
+
+    #[test]
+    fn test_fake_credential_generator() {
```

Comment:
This can return `Result<(), WebauthnError>`, that way you don't need `unwrap()` everywhere.

Reply:
Unwraps tell us which line failed though.